### DISCUSSION TITLE
Which strategy is used for negative sampling??

### DISCUSSION THREAD

**Discussion by ElifEceErdem** [RUCAIBox/RecBole#1833] [Q&A] [answered]
Hello,  
I have a question about negative sampling. In Recbole documentation, it says: "we will set positive label for all the observed interactions, and select some negative samples from the unobserved interactions according to a **certain strategy**." Can you explain the strategy this library uses while implementing the negative sapling for labeling the data?

Thanks in advance.

**Comment by Sherry-XLL** [comment, marked as answer]
Hello @tripledoubleE, thanks for your attention to RecBole!

In RecBole, we support three commonly used negative sampling methods, including random negative sampling (RNS), popularity-biased negative sampling (PNS) and dynamic negative sampling (DNS).

[Training Settings — RecBole 1.1.1 documentation](https://recbole.io/docs/user_guide/config/training_settings.html)

RecBole samples `num` negative items for each user in a batch. We first implement `get_used_ids` to obtain `used_ids`, which are composed of positive items. When negative sampling, we will not sample from these items. Then we implement `self.sampling` to sample negative items. If the distribution is uniform, the negative items will be sampled by `np.random.randint` in a uniform distribution. Else if the distribution is popularity, the sample distribution will be based on item popularity. And, `dynamic (bool)` decides whether we adopt dynamic negative sampling.

If you have any usage issues with our library, we suggest that you open a new [issue](https://github.com/RUCAIBox/RecBole/issues) instead of a discussion, so that we can better answer your questions.

**Reply by ElifEceErdem** [reply]
thank you so much
